lib/media.js

```diff
@@ -25,16 +32,81 @@ exports.init = function () {
         return
       }
 
-      // Homey.log('media search result', result.items[0])
       lastQuery = parsedQuery.searchQuery
-      lastResults = result.items.map(function (item) {
-        return exports.parseTrack(item)
+      // Build a list with video ids to get the durations for the videos in 1 API call
+      var idList = result.items.map(function (item) {
+        return item.id.videoId
+      }).join(',')
+
+      youtube('videos', {
+        part: 'id,snippet,contentDetails',
+        id: idList
+      }, function (err, result) {
+        if (err) {
+          callback(err)
+        } else {
+          var lastResults = result.items.map(function (item) {
+            return exports.parseTrack(item)
+          })
+          callback(null, lastResults.filter(function (item) { return item.duration < MAX_VIDEO_DURATION })) // Filter results that are longer than 2H because the MP3 API does not support this
+        }
       })
-      callback(null, lastResults)
     })
   })
 
   Homey.manager('media').on('play', exports.playVideoId)
+
+  // Sync playlists from youtube if the user has logged in
+  Homey.manager('media').on('getPlaylists', (data, callback) => {
+    console.log('Getting playlists')
+    oauth.getAccessToken(function (err, accessToken) {
+      if (err) {
+        callback(err)
+        return
+      }
+      // Call the api for playlists
+      http.json({
+        protocol: 'https:',
+        host: 'www.googleapis.com',
+        path: '/youtube/v3/playlists?part=id,snippet&mine=true&maxResults=' + MAX_PLAYLISTS + '&access_token=' + accessToken
+      }).then(function (response) {
+        if (response && response.items) {
+          var playlistRequests = response.items.map((item) => {
+            return new Promise((resolve) => {
+              getPlaylistItems(item.id, function (err, playlistItems) {
+                if (err) {
+                  callback(err, [])
```

Comment:
we should reject promise or `return` here. I recommend reject and handle rejection in `Promise.all` below.